compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -907,19 +907,21 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         did: DefId,
         item_segment: &hir::PathSegment<'_>,
     ) -> Ty<'tcx> {
+        let tcx = self.tcx();
         let args = self.ast_path_args_for_ty(span, did, item_segment);
-        let ty = self.tcx().at(span).type_of(did);
+        let ty = tcx.at(span).type_of(did);
 
-        if matches!(self.tcx().def_kind(did), DefKind::TyAlias)
-            && (ty.skip_binder().has_opaque_types() || self.tcx().features().lazy_type_alias)
+        if let DefKind::TyAlias = tcx.def_kind(did)
+            && (ty.skip_binder().has_opaque_types() || tcx.type_aliases_are_lazy(did.krate))
```

Comment:
Although I did not try it out, instead of invoking `tcx.type_aliases_are_lazy` here, I could alternatively invoke it when collecting the constrains from the generic arguments of weak projections. This should also fix #114468 but I haven't thought about the consequences yet. I don't like this as much as my current patch though as it doesn't address the points I raised in section “edition interoperability”.

Reply:
We could also store the laziness in `DefKind::TyAlias` as a boolean field. Do you think that would resolve the issues entirely?

Reply:
This would also be an option but it requires updating quite a few places. I've just pushed a commit so you can decide for yourself if you prefer it this way. If you approve of it, I'll squash the commits.

One slight advantage of `CrateRoot.type_aliases_are_lazy` is the fact that we can just throw it out if / once lazy type aliases get stabilized in edition 202X since we can simply do

```diff
-  type_aliases_are_lazy => { cdata.root.type_aliases_are_lazy }
+ type_aliases_are_lazy => { cdata.root.edition.at_least_rust_202x() }
```

Reply:
I like the way it looks now. We can always convert to another scheme, but I think encoding this information in the type alias DefKind matches how we encode similar other differences (e.g. mutable vs nonmutable statics)